### PR TITLE
Support migrating to processes by process name

### LINKED ISSUE #6313: Support migrating by process name
Perhaps there is an existing, better way of doing this, but I don't know any.  It is annoying to have to locate an appropriate process to migrate to (say, a 64-bit one) and then get its PID to pass to `migrate`, when most of the time you probably know exactly which process name you want to migrate to given that there is presumably a common process name (like vmtoolsd.exe) that suits your needs.  I don't see why `migrate` couldn't take a non-integer value (perhaps optionally) and search through `ps` for the right process.

### PR DESCRIPTION
**Pull request by jhart-r7**
Fixes #6313, assuming there isn't already a better way of doing this.

Existing functionality remains unchanged, but you can now pass a PID or process name as an explicit, named argument.  If you pass the PID as a named argument, it functions as if you just specified the PID alone (IOW, `migrate 12345` == `migrate -P 12345`), but if you pass a process name, it finds the PID of that process and migrates to it:

```
meterpreter > ps

Process List
============

 PID   PPID  Name               Arch  Session  User                           Path
 ---   ----  ----               ----  -------  ----                           ----
 0     0     [System Process]                                                 
 4     0     System             x64   0                                       
 100   516   svchost.exe        x64   0        NT AUTHORITY\NETWORK SERVICE   C:\Windows\System32\svchost.exe
 264   4     smss.exe           x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\smss.exe
 356   348   csrss.exe          x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\csrss.exe
 408   348   wininit.exe        x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\wininit.exe
 420   400   csrss.exe          x64   1        NT AUTHORITY\SYSTEM            C:\Windows\System32\csrss.exe
 468   400   winlogon.exe       x64   1        NT AUTHORITY\SYSTEM            C:\Windows\System32\winlogon.exe
 516   408   services.exe       x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\services.exe
 524   408   lsass.exe          x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\lsass.exe
 532   408   lsm.exe            x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\lsm.exe
 612   608   powershell.exe     x86   0        NT AUTHORITY\SYSTEM            C:\Windows\syswow64\WindowsPowerShell\v1.0\powershell.exe
 632   516   svchost.exe        x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\svchost.exe
 708   356   conhost.exe        x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\conhost.exe
 712   516   svchost.exe        x64   0        NT AUTHORITY\NETWORK SERVICE   C:\Windows\System32\svchost.exe
 724   1160  regedit.exe        x64   1        ESSENTIALS-11-P\Administrator  C:\Windows\regedit.exe
 792   516   svchost.exe        x64   0        NT AUTHORITY\LOCAL SERVICE     C:\Windows\System32\svchost.exe
 852   516   svchost.exe        x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\svchost.exe
 876   516   svchost.exe        x64   0        NT AUTHORITY\LOCAL SERVICE     C:\Windows\System32\svchost.exe
 900   516   svchost.exe        x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\svchost.exe
 944   516   svchost.exe        x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\svchost.exe
 1004  516   svchost.exe        x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\svchost.exe
 1144  852   dwm.exe            x64   1        ESSENTIALS-11-P\Administrator  C:\Windows\System32\dwm.exe
 1160  1136  explorer.exe       x64   1        ESSENTIALS-11-P\Administrator  C:\Windows\explorer.exe
 1224  516   spoolsv.exe        x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\spoolsv.exe
 1272  516   svchost.exe        x64   0        NT AUTHORITY\LOCAL SERVICE     C:\Windows\System32\svchost.exe
 1384  516   svchost.exe        x64   0        NT AUTHORITY\LOCAL SERVICE     C:\Windows\System32\svchost.exe
 1420  3056  csrss.exe          x64   3        NT AUTHORITY\SYSTEM            C:\Windows\System32\csrss.exe
 1444  516   svchost.exe        x64   0        NT AUTHORITY\LOCAL SERVICE     C:\Windows\System32\svchost.exe
 1528  1160  vmtoolsd.exe       x64   1        ESSENTIALS-11-P\Administrator  C:\Program Files\VMware\VMware Tools\vmtoolsd.exe
 1544  516   SeaPort.EXE        x86   0        NT AUTHORITY\SYSTEM            C:\Program Files (x86)\Microsoft\BingBar\SeaPort.EXE
 1580  516   taskhost.exe       x64   1        ESSENTIALS-11-P\Administrator  C:\Windows\System32\taskhost.exe
 1788  516   SearchIndexer.exe  x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\SearchIndexer.exe
 1812  2816  LogonUI.exe        x64   3        NT AUTHORITY\SYSTEM            C:\Windows\System32\LogonUI.exe
 1880  516   cygrunsrv.exe      x86   0        NT AUTHORITY\SYSTEM            C:\cygwin\bin\cygrunsrv.exe
 1932  356   conhost.exe        x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\conhost.exe
 1948  1916  sshd.exe           x86   0        NT AUTHORITY\SYSTEM            C:\cygwin\usr\sbin\sshd.exe
 2000  516   svchost.exe        x64   0        NT AUTHORITY\NETWORK SERVICE   C:\Windows\System32\svchost.exe
 2004  516   vmtoolsd.exe       x64   0        NT AUTHORITY\SYSTEM            C:\Program Files\VMware\VMware Tools\vmtoolsd.exe
 2040  516   WLIDSVC.EXE        x64   0        NT AUTHORITY\SYSTEM            C:\Program Files\Common Files\Microsoft Shared\Windows Live\WLIDSVC.EXE
 2100  1160  MSASCui.exe        x64   1        ESSENTIALS-11-P\Administrator  C:\Program Files\Windows Defender\MSASCui.exe
 2156  100   rdpclip.exe        x64   1        ESSENTIALS-11-P\Administrator  C:\Windows\System32\rdpclip.exe
 2200  2040  WLIDSVCM.EXE       x64   0        NT AUTHORITY\SYSTEM            C:\Program Files\Common Files\Microsoft Shared\Windows Live\WLIDSVCM.EXE
 2508  516   dllhost.exe        x64   0        NT AUTHORITY\SYSTEM            C:\Windows\System32\dllhost.exe
 2664  516   msdtc.exe          x64   0        NT AUTHORITY\NETWORK SERVICE   C:\Windows\System32\msdtc.exe
 2816  3056  winlogon.exe       x64   3        NT AUTHORITY\SYSTEM            C:\Windows\System32\winlogon.exe

meterpreter > migrate 2816
[*] Migrating from 612 to 2816...
[*] Migration completed successfully.
meterpreter > migrate -N vmtoolsd.exe
[*] Migrating from 2816 to 1528...
[*] Migration completed successfully.
meterpreter > migrate -P 2816
[*] Migrating from 1528 to 2816...
[*] Migration completed successfully.
meterpreter >
```

### CODE REVIEW

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/core.rb [hunk 1]
```diff
@@ -840,31 +842,39 @@ def cmd_migrate_help
...
-      return
-    end
-
+    pid = nil
     writable_dir = nil
     opts = {
       timeout: nil
     }
 
-    @@transport_opts.parse(args) do |opt, idx, val|
```
Comment (jhart-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/core.rb]: This was a bit of a bug but fortunately it didn't show itself -- the options for transport are a superset of those for migrate, so even though we were parsing options for `migrate` using the parser for `transport` it didn't matter.

Comment (OJ) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/core.rb]: Thanks Jon. I had a PR ready for this options parsing issue, but fur some
reason hadn't pushed it yet.

Did you know about "run migrate -n process.exe" ?

Cheers!

Comment (jhart-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/core.rb]: I do now.  I guess this PR simply adds the functionality to this `migrate` that the other already has.

Comment (OJ) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/core.rb]: Yup, cool :)

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/core.rb [hunk 2]
```diff
@@ -816,16 +816,18 @@ def cmd_transport(*args)
   end
 
   @@migrate_opts = Rex::Parser::Arguments.new(
-    '-p'  => [true,  'Writable path - Linux only (eg. /tmp).'],
-    '-t'  => [true,  'The number of seconds to wait for migration to finish (default: 60).'],
-    '-h'  => [false, 'Help menu.']
+    '-P' => [true, 'PID to migrate to.'],
```
Comment (OJ) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/core.rb]: Given that so many of us are now geared to `migrate <pid>` should we no default to this anyway and just handle cases where it's not given and look for a process name instead when `-N` is given?

My guess is that using `migrate -P <pid>` is going to feel like moving from Windows 7 to Windows 8 and not having a start menu any more. I'd rather see it behave as it did, but have the new parameter for string-based process names.

Thoughts?

Comment (jhart-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/core.rb]: That functionality remains unchanged.  You can still use `migrate <pid>`. 

I added the `-P <pid>` option for those who are OCD like me and would find it weird to use two different option styles (bare and non-bare) depending on what you are passing.

Comment (OJ) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/core.rb]: Ah good stuff. Thanks mate :+1: I should read further and not make assumptions :)

**Comment by wchen-r7** [comment]
Trying this out.

**Comment by wchen-r7** [comment]
This works for me (with ```migrate [pid]```, or ```migrate  -N [name]```, or ```migrate -P [pid]```).

**Pull request merged**